### PR TITLE
Added graf component testing

### PR DESCRIPTION
**Pull request by vasylvovkglo**

### CODE REVIEW

File: tests/features/common/actions/common.action.js [hunk 1]
```diff
@@ -40,6 +40,7 @@ const action = {
       await scrollToWebElement(driver, element)
     }
     const coordinates = await element.getRect()
+    console.log(coordinates)
```
Comment (dmytro-panchuk) [inline tests/features/common/actions/common.action.js]: console.log

File: tests/features/common/actions/graph.action.js [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+  getNamedFieldsGeometry
+} from './table.action'
+
+// import { DataFrame } from 'pandas-js'
+
+import numjs from 'numjs'
+import { expect } from 'chai'
+
+function differenciator(array0, array1, deviation) {
+  console.log(array0, array1)
```
Comment (dmytro-panchuk) [inline tests/features/common/actions/graph.action.js]: console.log

File: tests/features/step-definitions/steps.js [hunk 3]
```diff
@@ -61,6 +62,10 @@ import {
   typeSearchableValue,
   isContainsSubstringInSuggestedOptions
 } from '../common/actions/input-with-autocomplete.action'
+import {
+  // checkNodesConnections,
```
Comment (dmytro-panchuk) [inline tests/features/step-definitions/steps.js]: unnecessary comment

File: tests/features/common/actions/graph.action.js [hunk 4]
```diff
@@ -0,0 +1,250 @@
+import {
+  getRowsGeometry,
+  getFieldsGeometry,
+  getNamedRowsGeometry,
+  getNamedFieldsGeometry
+} from './table.action'
+
+// import { DataFrame } from 'pandas-js'
```
Comment (dmytro-panchuk) [inline tests/features/common/actions/graph.action.js]: unnecessary comment

File: tests/features/step-definitions/table.steps.js [hunk 5]
```diff
@@ -22,6 +22,8 @@ import {
   isContainsSubstringInColumnTooltipCells,
   isDatetimeCelsValueInRange,
   findRowIndexesByColumnTooltipsValue
+  // getRowsCoordinates,
+  // getfieldAttributeValues
```
Comment (dmytro-panchuk) [inline tests/features/step-definitions/table.steps.js]: unnecessary comments

File: tests/features/step-definitions/steps.js [hunk 6]
```diff
@@ -829,3 +834,27 @@ Then('select {string} with {string} value in breadcrumbs menu', async function(
     name
   )
 })
+
+Then(
+  'verify arrow lines position on {string} on {string} wizard',
+  async function(graphName, wizardName) {
+    // await this.driver.sleep(5000)
+    // await checkNodesConnections(this.driver, pageObjects[wizardName][graphName])
```
Comment (dmytro-panchuk) [inline tests/features/step-definitions/steps.js]: unnecessary comments

File: tests/features/common/actions/graph.action.js [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+import numjs from 'numjs'
+import { expect } from 'chai'
+
+function differenciator(array0, array1, deviation) {
+  console.log(array0, array1)
+  let diff = []
```
Comment (illia-prokopchuk) [inline tests/features/common/actions/graph.action.js]: You can use Const

File: tests/features/common/actions/graph.action.js [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+  for (let item0 of array0) {
+    let diffsMap = []
```
Comment (illia-prokopchuk) [inline tests/features/common/actions/graph.action.js]: You can use Const

File: tests/features/common/actions/graph.action.js [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+    })
+
+    const arrowsNodeLeftEdgesX = await arrowsNodeLeftEdges.map(item => item.x)
+    const topNodePointsX = await topNodePoints.map(item => item.x)
+    const leftEdgesConnections = differenciator(
+      arrowsNodeLeftEdgesX,
+      topNodePointsX,
+      maxConnectorSide
+    )
+    console.log('\n', leftEdgesConnections)
```
Comment (illia-prokopchuk) [inline tests/features/common/actions/graph.action.js]: remove console.log

**Review by illia-prokopchuk — CHANGES_REQUESTED** [review]

File: tests/features/common/actions/graph.action.js [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+      arrowsNodeLeftEdgesX,
+      nodeMidlePoints,
+      maxConnectorSide / 2
+    )
+    const righEdgesConnections = diffMapper(
+      arrowsNodeRightEdgesX,
+      nodeMidlePoints,
+      maxConnectorSide / 2
+    )
+
```
Comment (illia-prokopchuk) [inline tests/features/common/actions/graph.action.js:100]: remove empty line

File: tests/features/common/actions/graph.action.js [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    const startMarks = numjs.add(
+      numjs.multiply(startArrowsEdges, leftEdgesConnections),
+      numjs.multiply(startArrowsEdges, righEdgesConnections)
+    )
+    for (let i = 0; i < arrowsNodeRightEdgesX.length; i++) {
```
Comment (illia-prokopchuk) [inline tests/features/common/actions/graph.action.js:105]: add empty line before "for"

File: tests/features/common/actions/graph.action.js [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+      expect(numjs.sum(startMarks.slice([i]))).equal(true, 'Arrow not started')
+    }
+
+    const endMarks = numjs.add(
+      numjs.multiply(endArrowsEdges, leftEdgesConnections),
+      numjs.multiply(endArrowsEdges, righEdgesConnections)
+    )
```
Comment (illia-prokopchuk) [inline tests/features/common/actions/graph.action.js:112]: add empty line before "for"

File: tests/features/common/actions/table.action.js [hunk 13]
```diff
@@ -55,6 +58,7 @@ const action = {
     const rows = await getTableRows(driver, table)
```
Comment (illia-prokopchuk) [inline tests/features/common/actions/table.action.js:58]: add empty line after const

File: tests/features/common/actions/table.action.js [hunk 14]
```diff
@@ -55,6 +58,7 @@ const action = {
...
     expect(rows).not.equal(0)
     let flag = true
```
Comment (illia-prokopchuk) [inline tests/features/common/actions/table.action.js:59]: move variable declaration to 59 row, and add empty line before "expect"

File: tests/features/common/actions/table.action.js [hunk 15]
```diff
@@ -80,6 +85,7 @@ const action = {
     const rows = await getTableRows(driver, table)
     const arr = []
```
Comment (illia-prokopchuk) [inline tests/features/common/actions/table.action.js]: add empty line

File: tests/features/common/actions/table.action.js [hunk 16]
```diff
@@ -80,6 +85,7 @@ const action = {
...
     expect(rows).not.equal(0)
+
     for (let i = rows; i >= 1; i--) {
       await hoverComponent(driver, table.tableFields[column](i)['label'])
       const text = await getElementText(
```
Comment (illia-prokopchuk) [inline tests/features/common/actions/table.action.js:94]: move variable declaration to the top of loop, add empty line after variable declaration

File: tests/features/common/actions/table.action.js [hunk 17]
```diff
@@ -181,14 +191,16 @@ const action = {
     if (order === 'desc') {
       sortedArr = columnArray.reverse()
     }
+
     expect(sortedArr).equal(columnArray)
   },
   checkTableColumnValues: async function(driver, table, columnName, values) {
     const arr = await getColumnValues(driver, table, columnName)
-    expect(differenceWith(arr, values, isEqual).length).equal(0)
+    const diff = differenceWith(arr, values, isEqual)
```
Comment (illia-prokopchuk) [inline tests/features/common/actions/table.action.js:202]: add empty line below

File: tests/features/common/actions/table.action.js [hunk 18]
```diff
@@ -208,7 +220,60 @@ const action = {
...
+
+    return result
+  },
+  getRowsGeometry: async function(driver, table) {
+    const rowsNumber = await getTableRows(driver, table)
+    const result = []
+
+    for (let i = 1; i <= rowsNumber; i++) {
+      const rowRoot = await driver.findElement(table.rowRoot(i))
+      const position = await rowRoot.getRect()
```
Comment (illia-prokopchuk) [inline tests/features/common/actions/table.action.js:236]: add empty line below

File: tests/features/common/actions/table.action.js [hunk 19]
```diff
@@ -208,7 +220,60 @@ const action = {
...
+  getNamedRowsGeometry: async function(driver, table, name = 'name') {
+    const rowsNumber = await getTableRows(driver, table)
+    const result = []
+
+    for (let i = 1; i <= rowsNumber; i++) {
+      const rowRoot = await driver.findElement(table.rowRoot(i))
+      const rowName = await driver
+        .findElement(table.tableFields[name](i))
+        .getText()
+      const position = await rowRoot.getRect()
```
Comment (illia-prokopchuk) [inline tests/features/common/actions/table.action.js:264]: add empty line below

File: tests/features/common/components/graph.component.js [hunk 20]
```diff
@@ -0,0 +1,35 @@
+import { By } from 'selenium-webdriver'
+import { locatorBuilder } from '../../common-tools/common-tools'
+
+module.exports = function(graphStructure) {
+  const options = locatorBuilder`${0} ${1}`
+  const nodesTable = {
+    ...graphStructure.elements.workflowGrafNodesTable.structure
+  }
```
Comment (illia-prokopchuk) [inline tests/features/common/components/graph.component.js:8]: add empty line below

File: tests/features/common/components/graph.component.js [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+  nodesTable.root = options(
+    graphStructure.root,
+    graphStructure.elements.workflowGrafNodesTable.root
+  )
+
+  const grafConnections = {
```
Comment (illia-prokopchuk) [inline tests/features/common/components/graph.component.js:15]: Move variable declaration to the top of the function body

File: tests/features/common/page-objects/info-pane.po.js [hunk 22]
```diff
@@ -266,6 +267,64 @@ const featureSetsInfoPaneLabelsTable = {
...
+    offset: 1,
+    row: {
+      root: '.graph-pane__row',
+      fields: {
+        tab: ' .graph-pane__row-label'
+      }
+    }
+  }
+}
+
```
Comment (illia-prokopchuk) [inline tests/features/common/page-objects/info-pane.po.js:285]: delete empty line

File: tests/features/common/page-objects/jobs-and-workflows.po.js [hunk 23]
```diff
@@ -165,6 +166,48 @@ const workflowsMonitorTable = {
...
+            }
+          }
+        }
+      }
+    },
+    svg: '.react-flow__edges',
+    zoomPane: '.react-flow__nodes'
+  }
+}
+
```
Comment (illia-prokopchuk) [inline tests/features/common/page-objects/jobs-and-workflows.po.js:210]: delete empty line

File: tests/features/step-definitions/steps.js [hunk 24]
```diff
@@ -61,6 +62,7 @@ import {
   typeSearchableValue,
   isContainsSubstringInSuggestedOptions
 } from '../common/actions/input-with-autocomplete.action'
+import { checkNodesConnectionsNPandas } from '../common/actions/graph.action'
```
Comment (illia-prokopchuk) [inline tests/features/step-definitions/steps.js]: delete empty line

**Review by illia-prokopchuk — CHANGES_REQUESTED** [review]

**Pull request merged**